### PR TITLE
Optimize threads

### PR DESCRIPTION
**Pull request by oliver-batchelor**
I've optimized feature extraction and feature matching to make better use of the GPU by reading inputs and writing outputs in threads. 

* Makes feature extractors more fully use the GPU via. threaded output writer
* Speeds up NN feature matcher (on my machine a factor of 10)
* Multi-threading for cpu only (SIFT) extractor (relies on adding a scoped_gil_release to pycolmap too)

### CODE REVIEW

**Comment by Skydes** [comment]
Very nice, thank you! I don't have time to test in details for now but I left some high-level preliminary comments.

File: hloc/match_features.py [hunk 1]
```diff
@@ -90,74 +95,132 @@ def main(conf: Dict,
...
-            matches = pred['matches0'][0].cpu().short().numpy()
-            grp.create_dataset('matches0', data=matches)
-
-            if 'matching_scores0' in pred:
-                scores = pred['matching_scores0'][0].cpu().half().numpy()
-                grp.create_dataset('matching_scores0', data=scores)
-
-        skip_pairs.add(pair)
+    pairs = [pair for pair in set(pairs) 
+      if names_to_pair(*pair) not in skip_pairs]
```
Comment (Skydes) [inline hloc/match_features.py:203]: Here we need to also
1. check for existence in `skip_pairs` for both directions 0->1 and 1->0
2. Remove duplicated equivalent pairs

File: hloc/match_features.py [hunk 2]
```diff
@@ -90,74 +95,132 @@ def main(conf: Dict,
...
+
+    feature_pairs = FeaturesPairs(pairs, feature_path_q, feature_paths_refs)
+    loader = torch.utils.data.DataLoader(feature_pairs, num_workers=num_workers, batch_size=batch_size, shuffle=False, pin_memory=True)
```
[the remark this answers is not in the document]
Comment (oliver-batchelor) [reply hloc/match_features.py:206]: The NN matcher does (and is helpful) - re: SuperGlue you're right.

File: hloc/extract_features.py [hunk 3]
```diff
@@ -201,73 +204,104 @@ def __len__(self):
...
+          writer.put( (name, pred) )                     
+
+      process = WorkQueue( 
+        process_item = partial(process_image), 
+        num_threads=num_workers if device == 'cpu' else 1
+      )
+
+      with tqdm(loader) as pbar:
+        for data in pbar:
+          process.put(data)
```
[the remark this answers is not in the document]
Comment (oliver-batchelor) [reply hloc/extract_features.py:301]: Thanks - will check that out!

File: hloc/match_features.py [hunk 4]
```diff
@@ -90,74 +95,132 @@ def main(conf: Dict,
...
+
+    with h5py.File(str(match_path), 'a') as fd:
+
+      writer = WorkQueue(
+        process_item = partial(write_matches, fd=fd),
+        num_threads=num_workers)
+
+      with tqdm(smoothing=.1, total=len(feature_pairs)) as pbar:
+        for pairs, data in loader:
+            data = map_tensor(data, partial(torch.Tensor.to, device=device, dtype=torch.float16))
```
[the remark this answers is not in the document]
Comment (oliver-batchelor) [reply hloc/match_features.py:216]: I will check that out - mostly I have been using it with torch.amp.autocast, so I will check with and without (maybe it should be a an argument ... e.g. use_autocast or the like?)

File: hloc/match_features.py (hunk 1 above)
Comment (oliver-batchelor) [reply hloc/match_features.py:203]: Ah ha, thanks.

Comment (Skydes) [reply hloc/match_features.py:203]: I actually added this in PR https://github.com/cvg/Hierarchical-Localization/pull/159, I will merge it in the next days

**Comment by Skydes** [comment]
Cleaner implementation in https://github.com/cvg/Hierarchical-Localization/pull/242

**Pull request closed without merging**